File: c++/105. 从前序与中序遍历序列构造二叉树.cpp
#include<iostream>
#include<vector>
#include<unordered_map>
#include<stack>
using namespace std;
struct TreeNode {
	int val;
	TreeNode* left = nullptr;
	TreeNode* right = nullptr;
	TreeNode(int _val) :val(_val), left(nullptr), right(nullptr) {};
};
class Solution_105 {
public:
    int pre_index = 0;
    int post_size;
    unordered_map<int, int> map;
    TreeNode* create_tree(int left, int right, vector<int>& preorder, vector<int>& inorder) {
        if (left > right) {
            return nullptr;
        }
        int _val = preorder[pre_index++];
        int pose = map[_val];
        TreeNode* root = new TreeNode(_val);
        root->left = create_tree(left, pose - 1, preorder, inorder);
        root->right = create_tree(pose + 1, right, preorder, inorder);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        post_size = preorder.size();
        if (post_size < 1) return nullptr;
        for (auto& s : inorder) {
            map[s] = pre_index++;
        }
        pre_index = 0;
        return create_tree(0, post_size - 1, preorder, inorder);
    }
	void show_tree_inorder(TreeNode* root) {
		stack<TreeNode*> st;
		st.push(root);
		while (!st.empty()) {
			TreeNode* cur = st.top();
			cout << cur->val << "  ";
			st.pop();
			if (cur->right) {
				st.push(cur->right);
			}
			if (cur->left) {
				st.push(cur->left);
			}
		}
	}
};

File: python/105. 从前序与中序遍历序列构造二叉树.py
class TreeNode:
    def __init__(self,_val,left=None,right=None):
        self.val=_val
        self.left=left
        self.right=right
class Solution_105:
    def build_tree(self,preorder,inorder):
        def create_tree(l,r):
            if l>r:
                return None
            _val=preorder[0]
            del preorder[0]
            # preorder=preorder[i]
            index=index_map[_val]
            root=TreeNode(_val)
            root.left=create_tree(l,index-1)
            root.right=create_tree(index+1,r)
            return root
        index_map={val:index for index,val in enumerate(inorder)}
        return create_tree(0,len(preorder)-1)
    def show_tree_preorder(self,root):
        if(not root):
            return None
        print(root.val,"  ")
        self.show_tree_preorder(root.left)
        self.show_tree_preorder(root.right)
if __name__ == '__main__':
    preorder=[3,9,20,15,7]
    inorder=[9,3,15,20,7]
    s_105=Solution_105()
    root=s_105.build_tree(preorder,inorder)
    s_105.show_tree_preorder(root)

